cli/src/cmd/cast/estimate.rs

```diff
@@ -0,0 +1,102 @@
+// cast estimate subcommands
+use crate::opts::{cast::parse_name_or_address, EthereumOpts};
+use crate::utils::parse_ether_value;
+use cast::{Cast, TxBuilder};
+use clap::Parser;
+use ethers::{
+    providers::{Middleware, Provider},
+    types::{NameOrAddress, U256},
+};
+use foundry_config::{Chain, Config};
+
+#[derive(Debug, Parser)]
+pub struct EstimateArgs {
+    #[clap(help = "The destination of the transaction.", parse(try_from_str = parse_name_or_address), value_name = "TO")]
+    to: Option<NameOrAddress>,
+    #[clap(help = "The signature of the function to call.", value_name = "SIG")]
+    sig: Option<String>,
+    #[clap(help = "The arguments of the function to call.", value_name = "ARGS")]
+    args: Vec<String>,
+    #[clap(
+        long,
+        help = "Ether to send in the transaction.",
+        long_help = r#"Ether to send in the transaction, either specified in wei, or as a string with a unit type.
+
+Examples: 1ether, 10gwei, 0.01ether"#,
+        parse(try_from_str = parse_ether_value),
+        value_name = "VALUE"
+    )]
+    value: Option<U256>,
+    #[clap(flatten)]
+    // TODO: We only need RPC URL and Etherscan API key here.
+    eth: EthereumOpts,
+    #[clap(subcommand)]
+    command: Option<EstimateSubcommands>,
+}
+
+#[derive(Debug, Parser)]
+pub enum EstimateSubcommands {
+    #[clap(name = "--create", about = "Estimate gas cost to deploy a smart contract")]
+    Create {
+        #[clap(help = "Bytecode of contract.", value_name = "CODE")]
+        code: String,
+        #[clap(help = "The signature of the function to call.", value_name = "SIG")]
+        sig: Option<String>,
```

Comment:
Should this always be `constructor(address, uint256)`, i.e. "constructor" should be the sig's name for non-empty constructors?

Reply:
Not sure I follow.

`sig` could be more aptly named `constructor` for sure. But the reason it's optional is to allow for `DATA` to include constructor and args bytecode. Then `sig` and `args` shouldn't be passed in. 

iirc that's the behavior in `seth`  as well.

Reply:
> Then sig and args shouldn't be passed in.

let's add clap `conflicts_with = "<name>"` attributes then, requires `name = "..."` I blieve

Reply:
@mattsse just want to clarify to make sure I understand. There's three ways `cast estimate --create` could be used. 

If the contract bytecode is `<CODE>` and we have a costructor signature `constructor(address, uint256)` and args `X` &  `1`.  

We could do: 
1. `cast estimate --create <CODE> constructor(address, uint256) X 1`

Alternatively, we could pass in the bytecode of `X` and `1` into the `<DATA>` field. 
2. `cast estimate --create <CODE> DATA`

Finally, we could append <CODE> and the generated bytecode of `X` and `1` and pass that in as the code argument. 
3. `cast estimate --create <CODE>`.

If I'm understanding the `conflicts_with` attribute, it's based on the existence of a  parameter name. In this case, `<SIG>` & `<ARGS>` are optional based on the data passed into `<CODE>` (i.e. if the bytecode passed in contains the signature & argumetns encoded). They don't technically conflict with the existance any other argument -- we can use all three of them together -- so we wouldn't be able to use `conflicts_with` as a attribute. 

Let me know what you think, and if I'm understanding `conflicts_with` correctly.